server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java

```diff
@@ -126,8 +126,15 @@ public Builder(String name, IndexAnalyzers indexAnalyzers, ScriptCompiler script
 
             this.dimension = Parameter.boolParam("dimension", false, m -> toType(m).dimension, false)
                 .setValidator(v -> {
+                    if (v && (indexed.getValue() == false || hasDocValues.getValue() == false)) {
+                        throw new IllegalArgumentException(
+                            "Field [dimension] requires that [" + indexed.name + "] and [" + hasDocValues.name + "] are true"
+                        );
+                    }
                     if (v && ignoreAbove.getValue() < ignoreAbove.getDefaultValue()) {
-                        throw new IllegalArgumentException("Field [ignore_above] cannot be set in conjunction with field [dimension]");
+                        throw new IllegalArgumentException(
+                            "Field [" + ignoreAbove.name + "] cannot be set in conjunction with field [dimension]"
+                        );
                     }
```

Comment:
I think we should also refuse when there is a `normalizer` - it'd make it pretty complex to reason about the normalized values.